### PR TITLE
Universal Layers

### PR DESCRIPTION
**Pull request by Lyfts**
~~This had to be based on #4 and will be drafted until that is merged.~~

Aims to make it significantly easier to create layers by making the rendering as consistent as can be. As it is two very different mods there are still quirks that can't easily be worked around, but those quirks are completely irrelevant for the majority of layers.

### CODE REVIEW

File: src/main/java/com/gtnewhorizons/navigator/api/journeymap/render/JMInteractableLayerRenderer.java [hunk 1]
```diff
@@ -8,11 +8,12 @@
 import com.gtnewhorizons.navigator.api.NavigatorApi;
 import com.gtnewhorizons.navigator.api.journeymap.drawsteps.JMInteractableStep;
 import com.gtnewhorizons.navigator.api.journeymap.drawsteps.JMRenderStep;
+import com.gtnewhorizons.navigator.api.model.layers.InteractableLayer;
 import com.gtnewhorizons.navigator.api.model.layers.InteractableLayerManager;
 import com.gtnewhorizons.navigator.api.model.steps.RenderStep;
 import com.gtnewhorizons.navigator.api.util.Util;
 
-public abstract class JMInteractableLayerRenderer extends JMLayerRenderer {
+public abstract class JMInteractableLayerRenderer extends JMLayerRenderer implements InteractableLayer {
```
Comment (OneEyeMaker) [inline src/main/java/com/gtnewhorizons/navigator/api/journeymap/render/JMInteractableLayerRenderer.java:18]: Please add missing `@Override' annotations to mark implemented/overridden methods.

File: src/main/java/com/gtnewhorizons/navigator/api/model/layers/UniversalInteractableRenderer.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+import net.minecraft.client.gui.FontRenderer;
+
+import com.gtnewhorizons.navigator.api.NavigatorApi;
+import com.gtnewhorizons.navigator.api.model.steps.RenderStep;
+import com.gtnewhorizons.navigator.api.model.steps.UniversalInteractableStep;
+import com.gtnewhorizons.navigator.api.util.ClickPos;
+import com.gtnewhorizons.navigator.api.util.Util;
+
+public class UniversalInteractableRenderer extends UniversalLayerRenderer implements InteractableLayer {
```
Comment (OneEyeMaker) [inline src/main/java/com/gtnewhorizons/navigator/api/model/layers/UniversalInteractableRenderer.java:18]: Same

File: src/main/java/com/gtnewhorizons/navigator/api/journeymap/render/JMInteractableLayerRenderer.java (hunk 1 above)
Comment (OneEyeMaker) [reply src/main/java/com/gtnewhorizons/navigator/api/journeymap/render/JMInteractableLayerRenderer.java:18]: Also, annotations for nullability would be appreciated.

File: src/main/java/com/gtnewhorizons/navigator/api/util/DrawUtils.java [hunk 3]
```diff
@@ -163,30 +239,32 @@ public static void drawLabel(String text, double textX, double textY, int fontCo
...
         GL11.glPopMatrix();
     }
 
     public static float[] floats(int rgb) {
         return new float[] { (float) (rgb >> 16 & 255) / 255.0F, (float) (rgb >> 8 & 255) / 255.0F,
             (float) (rgb & 255) / 255.0F };
     }
+
+    public static int[] ints(int rgb, int alpha) {
+        return new int[] { rgb >> 16 & 255, rgb >> 8 & 255, rgb & 255, alpha & 255 };
```
Comment (OneEyeMaker) [inline src/main/java/com/gtnewhorizons/navigator/api/util/DrawUtils.java]: nitpick: Visually mark precedence of operators.
```suggestion
        return new int[] { (rgb >> 16) & 255, (rgb >> 8) & 255, rgb & 255, alpha & 255 };
```

File: src/main/java/com/gtnewhorizons/navigator/mixins/late/xaerosworldmap/GuiMapMixin.java [hunk 4]
```diff
@@ -130,6 +137,13 @@ protected GuiMapMixin(GuiScreen parent, GuiScreen escape) {
         }
 
         for (LayerRenderer layer : NavigatorApi.getActiveRenderersByPriority(XaeroWorldMap)) {
+            if (layer instanceof UniversalLayerRenderer universalLayerRenderer) {
+                for (XaeroRenderStep step : universalLayerRenderer.getRenderSteps()) {
```
Comment (OneEyeMaker) [inline src/main/java/com/gtnewhorizons/navigator/mixins/late/xaerosworldmap/GuiMapMixin.java:141]: There should be `instanceof` check like in `MinimapRendererMixin`, shouldn't it?

File: src/main/java/com/gtnewhorizons/navigator/api/model/steps/UniversalInteractableStep.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+import com.gtnewhorizons.navigator.api.journeymap.drawsteps.JMInteractableStep;
+import com.gtnewhorizons.navigator.api.model.locations.IWaypointAndLocationProvider;
+import com.gtnewhorizons.navigator.api.xaero.rendersteps.XaeroInteractableStep;
+
+import cpw.mods.fml.common.Optional;
+
+@Optional.Interface(
+    iface = "com.gtnewhorizons.navigator.api.journeymap.drawsteps.JMInteractableStep",
+    modid = "journeymap")
```
Comment (OneEyeMaker) [inline src/main/java/com/gtnewhorizons/navigator/api/model/steps/UniversalInteractableStep.java:14]: Should `@Optional` annotation be added for type for Xaero's Map, shouldn't it?

File: src/main/java/com/gtnewhorizons/navigator/api/model/steps/UniversalRenderStep.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+import com.gtnewhorizons.navigator.api.NavigatorApi;
+import com.gtnewhorizons.navigator.api.journeymap.drawsteps.JMRenderStep;
+import com.gtnewhorizons.navigator.api.model.locations.ILocationProvider;
+import com.gtnewhorizons.navigator.api.util.DrawUtils;
+import com.gtnewhorizons.navigator.api.xaero.rendersteps.XaeroRenderStep;
+
+import cpw.mods.fml.common.Optional;
+import journeymap.client.render.map.GridRenderer;
+
+@Optional.Interface(iface = "com.gtnewhorizons.navigator.api.journeymap.drawsteps.JMRenderStep", modid = "journeymap")
```
Comment (OneEyeMaker) [inline src/main/java/com/gtnewhorizons/navigator/api/model/steps/UniversalRenderStep.java:20]: Same question.

**Review by OneEyeMaker — CHANGES_REQUESTED** [review]

File: src/main/java/com/gtnewhorizons/navigator/api/model/steps/UniversalInteractableStep.java (hunk 5 above)
Comment (Lyfts) [reply src/main/java/com/gtnewhorizons/navigator/api/model/steps/UniversalInteractableStep.java:14]: The problem is that the JMRenderSteps extend the Journeymap DrawStep while the Xaero ones don't extend anything outside of this mod.

File: src/main/java/com/gtnewhorizons/navigator/api/journeymap/render/JMInteractableLayerRenderer.java (hunk 1 above)
Comment (Lyfts) [reply src/main/java/com/gtnewhorizons/navigator/api/journeymap/render/JMInteractableLayerRenderer.java:18]: What do you believe should be annotated as Nullable in that class?

File: src/main/java/com/gtnewhorizons/navigator/api/model/steps/UniversalInteractableStep.java (hunk 5 above)
Comment (OneEyeMaker) [reply src/main/java/com/gtnewhorizons/navigator/api/model/steps/UniversalInteractableStep.java:14]: Perhaps such an optional interface should be added to Xaero's map-related types to allow such access to map mod internals *in the future*. But I leave this design choice up to you.

File: src/main/java/com/gtnewhorizons/navigator/mixins/late/xaerosworldmap/GuiMapMixin.java (hunk 4 above)
Comment (Lyfts) [reply src/main/java/com/gtnewhorizons/navigator/mixins/late/xaerosworldmap/GuiMapMixin.java:141]: No it's not really necessary, no clue why I used it for the minimap though.

File: src/main/java/com/gtnewhorizons/navigator/api/model/steps/UniversalInteractableStep.java (hunk 5 above)
Comment (Lyfts) [reply src/main/java/com/gtnewhorizons/navigator/api/model/steps/UniversalInteractableStep.java:14]: As someone that have spent a ton of time dealing with Xaero code, I can say with 100% certainty that it is definitely not desired by anyone. We are way better off designing things inhouse and forcing Xaero to work with it. (I can't even begin to explain the absurdities of some of the Xaero code)

File: src/main/java/com/gtnewhorizons/navigator/api/model/layers/UniversalInteractableRenderer.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    private final ClickPos clickPos = new ClickPos();
+    protected InteractableLayerManager manager;
+    protected UniversalInteractableStep<?> hoveredRenderStep = null;
+    private Predicate<ClickPos> clickAction;
+    private IntPredicate keyPressAction;
+
+    public UniversalInteractableRenderer(InteractableLayerManager manager) {
```
Comment (OneEyeMaker) [inline src/main/java/com/gtnewhorizons/navigator/api/model/layers/UniversalInteractableRenderer.java]:
```suggestion
    public UniversalInteractableRenderer(@Nonnnull InteractableLayerManager manager) {
```

File: src/main/java/com/gtnewhorizons/navigator/api/model/layers/UniversalInteractableRenderer.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+    public UniversalInteractableRenderer withClickAction(Predicate<ClickPos> action) {
+        this.clickAction = action;
+        return this;
+    }
+
+    public UniversalInteractableRenderer withKeyPressAction(IntPredicate keyPressAction) {
+        this.keyPressAction = keyPressAction;
+        return this;
+    }
```
Comment (OneEyeMaker) [inline src/main/java/com/gtnewhorizons/navigator/api/model/layers/UniversalInteractableRenderer.java]: Nullability annotation for arguments is missing (I suggest adding `@NonNull`).

File: src/main/java/com/gtnewhorizons/navigator/api/journeymap/render/JMInteractableLayerRenderer.java (hunk 1 above)
Comment (OneEyeMaker) [reply src/main/java/com/gtnewhorizons/navigator/api/journeymap/render/JMInteractableLayerRenderer.java:18]: `@Nullable` field (`hoveredDrawState`) and `@Nonnull` constructor argument.

File: src/main/java/com/gtnewhorizons/navigator/api/model/layers/UniversalInteractableRenderer.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+import com.gtnewhorizons.navigator.api.util.ClickPos;
+import com.gtnewhorizons.navigator.api.util.Util;
+
+public class UniversalInteractableRenderer extends UniversalLayerRenderer implements InteractableLayer {
+
+    private final ClickPos clickPos = new ClickPos();
+    protected InteractableLayerManager manager;
+    protected UniversalInteractableStep<?> hoveredRenderStep = null;
+    private Predicate<ClickPos> clickAction;
+    private IntPredicate keyPressAction;
```
Comment (OneEyeMaker) [inline src/main/java/com/gtnewhorizons/navigator/api/model/layers/UniversalInteractableRenderer.java:24]: These fields should be `@Nullable` (if you're going to annotate fields).

**Review by OneEyeMaker — CHANGES_REQUESTED** [review]

File: src/main/java/com/gtnewhorizons/navigator/api/journeymap/render/JMInteractableLayerRenderer.java (hunk 1 above)
Comment (Lyfts) [reply src/main/java/com/gtnewhorizons/navigator/api/journeymap/render/JMInteractableLayerRenderer.java:18]: I have never annotated a field as either nullable or nonnull and really don't see the need for it either. You do have a point with the constructor arg though

**Review by OneEyeMaker — APPROVED** [review]

**Pull request merged**